src/thermostats/TemperatureObserver.cpp

```diff
@@ -0,0 +1,150 @@
+    for(auto& [region, region_cache] : _region_data) {
+        regions.emplace_back(region.low, region.high);
+    }
+}
+
+double TemperatureObserver::getTemperature(std::size_t index) {
+    mardyn_assert(index < _region_data.size() && index >= 0);
+    return _region_data[index].second.current_temp;
+}
+
+double TemperatureObserver::measureTemp(const std::array<double, 3> &low, const std::array<double, 3> &high,
+                                   ParticleContainer *particleContainer, BinData& binData) {
+    std::array<double, 3> v_mean = computeMeanVelocityStep(low, high, particleContainer, binData);
+    double m = _simulation.getEnsemble()->getComponent(0)->m();
+    double v = 0;
+    std::size_t n = 0;
+#if defined(_OPENMP)
+#pragma omp parallel reduction(+:v, n)
+#endif
+    for (auto it = particleContainer->regionIterator(std::data(low), std::data(high),
+                                                     ParticleIterator::ALL_CELLS); it.isValid(); ++it) {
+        for(int d = 0; d < 3; d++) {
+            v += std::pow(it->v(d) - v_mean[d], 2);
+        }
+        n += 1;
+    }
+
+    return v * m / (3 * static_cast<double>(n));
+}
+
+std::array<double, 3> TemperatureObserver::computeMeanVelocityStep(const std::array<double, 3> &low, const std::array<double, 3> &high,
+                                                              ParticleContainer *particleContainer, BinData& binData) {
+    //compute mean velocity per dim using SMC method, see: Karimian et al. 2011
+    std::array<double, 3> v_avg{0};
+    double* v_raw = std::data(v_avg);
+    std::size_t count = 0;
+#if defined(_OPENMP)
+#pragma omp parallel reduction(+:v_raw[:3], count)
+#endif
+    for (auto it = particleContainer->regionIterator(std::data(low), std::data(high),
+                                                     ParticleIterator::ALL_CELLS); it.isValid(); ++it) {
+        for(int d = 0; d < 3; d++) {
+            v_raw[d] += it->v(d);
+        }
+        count += 1;
+    }
+    //TODO check this
+    if(count == 0) count = 1;
+    binData.mol_counts.insert(count);
+    binData.velocities.insert(v_avg);
+
+    //compute CAM average first
+    {
+        std::size_t steps = binData.mol_counts.size();
+        auto d = static_cast<double>(steps<=1?1:steps-1);
+
+        std::array<double, 3> v_cam{ 0 };
+        for(std::size_t i = 0; i < steps; i++) {
+            const auto& vec = binData.velocities.get(i);
+            for(int dim = 0; dim < 3; dim++) {
+                v_cam[dim] += vec[dim];
+            }
+        }
+        for(int dim = 0; dim < 3; dim++) {
+            v_cam[dim] /= d;
+        }
+
+        std::size_t nks = 0;
+        for(std::size_t i = 0; i < steps; i++) {
+            nks += binData.mol_counts.get(i);
+        }
+        double nks_d = static_cast<double>(nks) / d;
+
+        for(int dim = 0; dim < 3; dim++) {
+            v_cam[dim] /= nks_d;
+        }
+        binData.v_cam.insert(v_cam);
+    }
+
+    //compute SAM over CAM
```

Comment:
What is SAM

Reply:
This is an averaging method, see 10.1002/fld.2487 and 10.1016/j.jcp.2003.10.021